submissions/xT0PERx/OOP exercise/index.js

```diff
@@ -0,0 +1,84 @@
+class Mammal {
+  constructor(name, gender, saying) {
+    this.name = name;
+    this.gender = gender;
+    this.saying = saying;
+  }
+  toString() {
+    return `${this.name}; ${this.gender}; ${this.saying}`;
```

Comment:
There is a pattern. Can you optimize this fragment employing `Array` methods?
Here and everywhere across all classes.

Reply:
Unfortunately, the possibility of simplifying the code using array methods does not occur to me. Perhaps you could somehow lead me to this idea? I looked at other works and found this approach.
return [this.species, this.name, this.gender, this.age] .join (',');
But I cannot understand why he is better? Maybe that's not what you mean.

Reply:
It is not about "better".
This is quite a simple use case. If you can spot and resolve the pattern in a simpler case then the chances you will spot in more complicated cases are higher. If you don't learn it now you will struggle later.
This may help you to avoid big blocks of template literals in e.g. Friends app and other, and also t oavoid repetitions of bigger blocks of code.
Do you want to master this skill?

Reply:
Of course I want, my task is to change my occupation. And while there is an opportunity to learn from me and from you to teach - I would like to use it.

Reply:
So what are your own on thoughts on the difference between using a template literal and `Array` methods?
Imagine you need to add a dozen properties more to print or to reorder the properties for output.

Reply:
To change the order of properties for display and to add a large number of elements, of course, it is better to use an array and its methods.
#### So I should add an array with elements and then reuse them? The way I see it:

> toString() {
    return [this.name, this.gender, this.saying].join(';');
  }

#### Then the Animal class will already have:

> toString() {
    return super.toString()+';'+[this.species, this.legs].join(';');
  }

That is, in the future, if I need to add any more features, I just add them to the array?
#### Maybe I misunderstood something?

Reply:
@xT0PERx 

The 2nd fragment also has a pattern thus optimizable.
```
toString() {
  return super.toString()+';'+[this.species, this.legs].join(';');
}
```

```
toString() {
  return [super.toString(), this.species, this.legs].join(';');
}
```

Reply:
Yet another perspective on the problem to reflect on

In
```
toString() {
  return super.toString()+';'+[this.species, this.legs].join(';');
}
```
if replace `super.toString()` speculatevely then the entire fragment look like the following
```
[this.name, this.gender, this.saying].join(';') + ';' + [this.species, this.legs].join(';')
```
Can you see the recurrent nature of the problem?

Reply:
With each task, I think that I am no longer a completely newbie, but with each review I am convinced of my inexperience. Thanks for your detailed comments.

Reply:
Learning is the endless process. Stretching and challenging is the way to learn.
Mentors challenge students one or two steps at a time to create stretching thus keeping the momentum for learning.